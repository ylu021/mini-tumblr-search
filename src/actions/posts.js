export const types = {
  ERROR_REQUESTING_POSTS: 'ERROR_REQUESTING_POSTS',
  RECEIVED_POSTS: 'RECEIVED_POSTS',
  REQUESTING_POSTS: 'REQUESTING_POSTS'
};

export const errorRequestingPosts = (err) => ({
  payload: {
    err
  },
  type: types.ERROR_REQUESTING_POSTS
});

export const receivedPosts = (posts) => ({
  payload: {
    posts
  },
  type: types.RECEIVED_POSTS
});

export const requestingPosts = (query) => ({
  payload: {
    query
  },
  type: types.REQUESTING_POSTS
});
