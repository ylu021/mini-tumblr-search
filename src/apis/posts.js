import path from 'path';

export const searchImg = async (query) => {
  const token = process.env.ACCESS_TOKEN;

  const endpoint = `https://api.tumblr.com/v2/tagged?tag=${query}&api_key=${token}`;
  const response = await fetch(endpoint);
  const data = await response.json();

  if (response.status < 200 || response.status >= 400) {
    const err = new Error(data && data.message);

    err.response = response;
    throw err;
  }

  const filteredOnlyPhotos = data.response.filter(
    ({photos}) => photos
  ).map(
    // eslint-disable-next-line id-match
    ({id, short_url, photos}) => ({
      id,
      url: short_url,
      image: photos
    })
  );

  return filteredOnlyPhotos;
};
