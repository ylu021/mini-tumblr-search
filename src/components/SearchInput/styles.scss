.searchinput {
  padding: 1rem;
  border-radius: 0.5rem 0 0 0.5rem;
  border: 0.15rem solid white;
  color: #333;
  font-family: inherit;
  font-size: large;
}

.title {
  line-height: 2;
}
