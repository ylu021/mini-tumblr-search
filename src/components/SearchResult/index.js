import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

const SearchResult = ({posts, fetching}) => {
  if (fetching) {
    return <p>Loading</p>;
  }

  if (!posts) {
    return <p />;
  }

  return (
    <div>
      <h1>Image result from Tumblr</h1>
      <ul>
        {
          posts.map(({id, image, url}) =>
            <li key={id}>
              <img src={image} />
              <a
                href={url}
                target='_blank'
              >
                {'source'}
              </a>
            </li>
          )
        }
      </ul>
    </div>
  );
};

SearchResult.propTypes = {
  fetching: PropTypes.bool.isRequired,
  posts: PropTypes.array
};

SearchResult.defaultProps = {
  posts: null
};

const mapStateToProps = (state) => ({
  fetching: state.posts.fetching,
  posts: state.posts.posts
});

export {SearchResult as SearchResultDummy};
export default connect(mapStateToProps)(SearchResult);
