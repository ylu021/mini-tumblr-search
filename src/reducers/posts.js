import {types} from '../actions/posts';

const {
  ERROR_REQUESTING_POSTS,
  RECEIVED_POSTS,
  REQUESTING_POSTS
} = types;

export const initialState = {
  err: null,
  fetching: false,
  posts: null
};

const posts = (state = initialState, action) => {
  switch (action.type) {
  case ERROR_REQUESTING_POSTS:
    return {
      ...state,
      err: action.payload.err
    };
  case RECEIVED_POSTS:
    return {
      ...state,
      fetching: false,
      posts: action.payload.posts
    };
  case REQUESTING_POSTS:
    return {
      ...state,
      fetching: true,
      posts: null
    };
  default:
    return state;
  }
};

export default posts;
