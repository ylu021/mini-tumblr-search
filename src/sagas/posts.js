import {call, fork, put, takeLatest} from 'redux-saga/effects';
import {searchImg} from '../apis/posts';
import {
  errorRequestingPosts,
  receivedPosts,
  types
} from '../actions/posts';

export const fetchFromTumblr = function *(action) {
  try {
    const {payload: {query}} = action;
    const receivedImgs = yield call(searchImg, query);

    yield put(receivedPosts(receivedImgs));
  } catch (error) {
    yield put(errorRequestingPosts(error));
  }
};

const watchRequestingPosts = function *() {
  yield takeLatest(types.REQUESTING_POSTS, fetchFromTumblr);
};

export default function *posts () {
  yield [
    fork(watchRequestingPosts)
  ];
}
