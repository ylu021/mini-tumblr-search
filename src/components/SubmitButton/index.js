import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {requestingPosts} from '../../actions/posts';
import styles from './styles.scss';

const SubmitButton = ({dispatch, fetching, query}) => {
  const handleSubmit = () => dispatch(requestingPosts(query));

  return (
    <button
      className={styles.submitbutton}
      disabled={fetching || !query}
      onClick={!fetching && handleSubmit}
    >
      Search
    </button>
  );
};

SubmitButton.propTypes = {
  dispatch: PropTypes.func.isRequired,
  fetching: PropTypes.bool.isRequired,
  query: PropTypes.string
};

SubmitButton.defaultProps = {
  query: null
};

const mapStateToProps = (state) => ({
  fetching: state.posts.fetching
});

export {SubmitButton as SubmitButtonDummy};
export default connect(mapStateToProps)(SubmitButton);
