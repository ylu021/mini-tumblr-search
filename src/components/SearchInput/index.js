import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {requestingPosts} from '../../actions/posts';
import SubmitButton from '../SubmitButton';

class SearchInput extends Component {
  state = {
    query: null
  };

  constructor (props) {
    super(props);
  }

  render () {
    const {query} = this.state;

    return (
      <form>
        <input
          name={'query'}
          onChange={
            (e) => this.setState({
              query: e.target.value
            })
          }
          type={'text'}
        />
        <SubmitButton query={query} />
      </form>
    );
  }
}

export {SearchInput as SearchInputDummy};
export default SearchInput;
