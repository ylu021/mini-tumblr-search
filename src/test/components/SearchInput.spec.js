import React, {Component} from 'react';
import {requestingPosts} from '../../actions/posts';
import {SubmitButtonDummy as SubmitButton} from '../../components/SubmitButton';
import {SearchInputDummy as SearchInput} from '../../components/SearchInput';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

const query = 'cat';

describe('<SearchInput />', () => {
  let component;

  beforeAll(() => {
    component = shallow(<SearchInput />);
  });
  it('should correctly renders nested component', () => {
    expect(component).toMatchSnapshot();
  });
});
