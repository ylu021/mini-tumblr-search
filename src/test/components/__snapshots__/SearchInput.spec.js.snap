// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> should correctly renders nested component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <form>
    <input
      name="query"
      onChange={[Function]}
      type="text"
    />
    <Connect(SubmitButton)
      query={null}
    />
  </form>,
  "nodes": Array [
    <form>
      <input
        name="query"
        onChange={[Function]}
        type="text"
      />
      <Connect(SubmitButton)
        query={null}
      />
    </form>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <SearchInput />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": SearchInput {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "query": null,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <form>
          <input
            name="query"
            onChange={[Function]}
            type="text"
          />
          <Connect(SubmitButton)
            query={null}
          />
        </form>,
        "_debugID": 2,
        "_renderedOutput": <form>
          <input
            name="query"
            onChange={[Function]}
            type="text"
          />
          <Connect(SubmitButton)
            query={null}
          />
        </form>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SearchInput />,
}
`;
